iot_verifier/src/gateway_updater.rs

```diff
@@ -0,0 +1,90 @@
+use crate::Settings;
+use chrono::Duration;
+use futures::stream::StreamExt;
+use helium_crypto::PublicKeyBinary;
+use iot_config::{
+    client::{Client as IotConfigClient, ClientError as IotConfigClientError},
+    gateway_info::{GatewayInfo, GatewayInfoResolver},
+};
+use std::collections::HashMap;
+use tokio::sync::watch;
+use tokio::time;
+
+pub type GatewayMap = HashMap<PublicKeyBinary, GatewayInfo>;
+pub type MessageSender = watch::Sender<GatewayMap>;
+pub type MessageReceiver = watch::Receiver<GatewayMap>;
+
+pub struct GatewayUpdater {
+    iot_config_client: IotConfigClient,
+    refresh_interval: Duration,
+    sender: MessageSender,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum GatewayUpdaterError {
+    #[error("gateway not found: {0}")]
+    GatewayNotFound(PublicKeyBinary),
+    #[error("error querying iot config service")]
+    IotConfigClient(#[from] IotConfigClientError),
+}
+
+impl GatewayUpdater {
+    pub async fn from_settings(
+        settings: &Settings,
+        mut iot_config_client: IotConfigClient,
+    ) -> Result<(MessageReceiver, Self), GatewayUpdaterError> {
+        let gateway_map = refresh_gateways(&mut iot_config_client).await?;
+        let (sender, receiver) = watch::channel(gateway_map);
+        Ok((
+            receiver,
+            Self {
+                iot_config_client,
+                refresh_interval: settings.gateway_refresh_interval(),
+                sender,
+            },
+        ))
+    }
+
+    pub async fn run(mut self, shutdown: &triggered::Listener) -> Result<(), GatewayUpdaterError> {
+        tracing::info!("starting gateway_updater");
+
+        let mut trigger_timer = time::interval(
+            self.refresh_interval
+                .to_std()
+                .expect("valid interval in seconds"),
+        );
+
+        loop {
+            if shutdown.is_triggered() {
+                tracing::info!("stopping gateway_updater");
+                return Ok(());
+            }
+
+            tokio::select! {
+                _ = trigger_timer.tick() => self.handle_refresh_tick().await?,
+                _ = shutdown.clone() => return Ok(()),
+            }
+        }
+    }
+
+    async fn handle_refresh_tick(&mut self) -> Result<(), GatewayUpdaterError> {
+        tracing::info!("handling refresh tick");
+        let updated_gateway_map = refresh_gateways(&mut self.iot_config_client).await?;
+        _ = self.sender.send(updated_gateway_map);
```

Comment:
would it be better to trigger this send error here with `self.sender.send(updated_gateway_map)?;`? i believe it's another case where that should only error in the event the receiver(s) have all been dropped

Reply:
addressed here: fdabbdfa32b12cc10f890d753fa15726fd7effd8